zebra-state/src/service/check/anchors.rs

```diff
@@ -331,8 +333,12 @@ pub(crate) fn block_sapling_orchard_anchors_refer_to_final_treestates(
     parent_chain: &Arc<Chain>,
     prepared: &PreparedBlock,
 ) -> Result<(), ValidateContextError> {
-    prepared.block.transactions.iter().enumerate().try_for_each(
-        |(tx_index_in_block, transaction)| {
+    prepared
+        .block
+        .transactions
+        .par_iter()
+        .enumerate()
+        .try_for_each(|(tx_index_in_block, transaction)| {
```

Comment:
This is mainly a disk operation, so it should be on the tokio thread, not a rayon thread.

Checking if a nullifier exists is usually cheap enough to be run serially, because there are usually 0-2 nullifiers revealed in a transaction, and RocksDB stores a hash index in-memory for nullifiers.

And remember, this request is already running in parallel with the proofs and the other async checks run by the transaction verifier.

So I don't think we should make this change without any evidence that it is actually slowing Zebra down:
https://github.com/ZcashFoundation/zebra/pull/5716#discussion_r1033033771

Reply:
> this request is already running in parallel with the proofs and the other async checks run by the transaction verifier.

This one's for the block, so it would be splitting up the work for a lot of transactions and it's used in the `validate_and_commit` method of `NonFinalizedState`.

I think it's different from the changes [here](https://github.com/ZcashFoundation/zebra/pull/4933/files) in that the work can be split up cheaply.

Reply:
By the way, the average number of transactions per block is 3, and one of those is a coinbase transaction:
https://bitinfocharts.com/zcash/

So the available amount of parallelism in a block is very low. And since rayon automatically adjusts parallel iterator splits to reduce overhead, there are no guarantees that each transaction will actually run in parallel, or that using multiple threads will be more efficient.

(It's different for proofs and note commitment tree updates, which can take around 1 second of CPU, so we can be doing a lot of them at the same time.)

Reply:
> By the way, the average number of transactions per block is 3

Oh. 🤦 Okay that makes sense. Thank you.

Reply:
It's not something I would expect anyone to know, I had to check it too 🙂